pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStoreTimerTask.java

```diff
@@ -0,0 +1,200 @@
+                            } else {
+                                //this judge the pendingAckMetadataEntry is can delete
+                                PositionImpl largestPosition = null;
+                                List<PendingAckMetadata> metadataList =
+                                        pendingAckMetadataEntry.getPendingAckMetadataList();
+                                for (int i = 0; i < metadataList.size(); i++) {
+                                    PendingAckMetadata pendingAckMetadata = metadataList.get(0);
+                                    if (largestPosition == null) {
+                                        largestPosition = PositionImpl.get(pendingAckMetadata.getLedgerId(),
+                                                pendingAckMetadata.getEntryId());
+                                    } else {
+                                        PositionImpl comparePosition = PositionImpl
+                                                .get(pendingAckMetadata.getLedgerId(),
+                                                        pendingAckMetadata.getEntryId());
+                                        if (largestPosition.compareTo(comparePosition) <  0) {
+                                            largestPosition = comparePosition;
+                                        }
+
+                                    }
+                                }
+                                if (largestPosition != null) {
+                                    handleAckCommon(largestPosition, nextPosition);
+                                }
+                            }
+                            break;
+                        case ABORT:
+                        case COMMIT:
+                            markDeletePosition = nextPosition;
+                            break;
+                        default:
+                            log.error("PendingAck timer task read illegal metadata state! {}",
+                                    pendingAckMetadataEntry.getPendingAckOp());
+                    }
+                } finally {
+                    entry.release();
+                    if (pendingAckMetadataEntry != null) {
+                        pendingAckMetadataEntry.recycle();
+                    }
+                    pendingAckMetadataEntryBuilder.recycle();
+                    stream.recycle();
+                }
+                // when markDeletePosition is not nextPosition, before markDeletePosition can delete
+                if (markDeletePosition != nextPosition) {
+                    this.managedCursor.markDelete(markDeletePosition);
+                    break;
+                }
+            }
+            this.timer.newTimeout(this, intervalTime, TimeUnit.MILLISECONDS);
+        } catch (Exception e) {
+            log.error("PendingAck timer task error!", e);
+            if ("Cursor was already closed".equals(e.getCause().getMessage())) {
```

Comment:
this is very error prone,
can we detect an instance of a specific class of exception ?